lib/spack/spack/bootstrap.py

```diff
@@ -837,3 +838,156 @@ def ensure_flake8_in_path_or_raise():
     """Ensure that flake8 is in the PATH or raise."""
     executable, root_spec = 'flake8', flake8_root_spec()
     return ensure_executables_in_path_or_raise([executable], abstract_spec=root_spec)
+
+
+def _missing_executable(name, purpose):
+    """Message to be printed if an executable is not found"""
+    msg = 'MISSING: "{0}" executable (required to {1})'
+    return msg.format(name, purpose)
+
+
+def _system_executable(exes, msg):
+    """Query availability of executables to be found on the system"""
+    if isinstance(exes, six.string_types):
+        exes = (exes,)
+    if spack.util.executable.which_string(*exes):
+        return True, None
+    return False, msg
+
+
+def _required_python_module(module, query_spec, msg):
+    """Query availability of required Python modules"""
+    if _python_import(module) or _try_import_from_store(module, query_spec):
+        return True, None
+    return False, msg
+
+
+def _required_executable(exes, query_spec, msg):
+    """Query availability of required executables"""
+    if isinstance(exes, six.string_types):
+        exes = (exes,)
+    if (spack.util.executable.which_string(*exes) or
+            _executables_in_store(exes, query_spec)):
+        return True, None
+    return False, msg
+
+
+def _core_requirements():
+    _core_system_exes = {
+        'make': _missing_executable('make', 'build software from sources'),
+        'patch': _missing_executable('patch', 'patch source code before building'),
+        'bash': _missing_executable('bash',  'use Spack compiler wrapper'),
+        'tar': _missing_executable('tar', 'manage code archives'),
+        'gzip': _missing_executable('gzip', 'compress/decompress code archives'),
+        'unzip': _missing_executable('unzip', 'compress/decompress code archives'),
+        'bzip2': _missing_executable('bzip2', 'compress/decompress code archives'),
+        'git': _missing_executable('git', 'fetch/manage git repositories')
+    }
+    if platform.system().lower() == 'linux':
+        _core_system_exes['xz'] = _missing_executable(
+            'xz', 'compress/decompress code archives'
+        )
+
+    # Executables that are not bootstrapped yet
+    result = [functools.partial(_system_executable, exe, msg)
+              for exe, msg in _core_system_exes.items()]
+    # Python modules
+    result += [
+        functools.partial(
+            _required_python_module, 'clingo', clingo_root_spec(),
+            'MISSING: "clingo" python module (required to concretize specs)'
+        )
```

Comment:
I think `_required_python_module` takes 3 arguments and I see 3 supplied args here, so I think you are using `functools.partial` as a convenience mechanism to call a supply a function object that calls a function with arguments. IMO that is not a straightforward use of `functools.partial` and I'd prefer just defining a `lambda` like

```
result += [lambda: _required_python_module(x, y, z)]
```

Also I'm not sure what the benefit of delayed execution is here: why not just run the function and store the results instead of saving the function for later?

Reply:
> IMO that is not a straightforward use of functools.partial and I'd prefer just defining a lambda like

The two things (`functools.partial` or lambda) are equivalent here. I was using this to get a list of callables without args which could return:
```
(found, error_message)
```
since I thought that would be clearer to read at call site. You're correct that `functools.partial` can be removed, and I did it in fe21e5a